### PR TITLE
Generic http wrapper

### PR DESCRIPTION
**Pull request by andreibuiciuc**
I implemented the http service in a generic manner. I also corrected two .spec.ts files that weren't updated after I renamed the respective services.

### CODE REVIEW

File: api/Controllers/UserController.cs [hunk 1]
```diff
@@ -55,15 +58,19 @@ public async Task<IActionResult> GetUserList()
...
         {
             if (await _userService.GetUserByEmail(user.Email) == null)
             {
                 var newUser = _mapper.Map<UserDto, User>(user);
                 var postUser = await _userService.PostUser(newUser);
-                return Ok("Your account was successfully created.");
+                return Ok(postUser);
             }
-            return BadRequest("User with given email already exists!");
+            return BadRequest("Email already in use.");
```
Comment (larisaChh) [inline api/Controllers/UserController.cs]: shouldnt this be a Conflict status error?

File: src/app/services/http.service.ts [hunk 2]
```diff
@@ -1,45 +1,68 @@
...
-  makeUserHttpRequest(requestName: string, requestBody?: any, ...pathParameters: string[]): Observable<any> {
-    const requestUrl: string = this.buildUrlPath(...pathParameters);
+  getAll<T>(endpoint: string): Observable<T[]> {
+    this.setRequestHeaders();
+    const requestUrl = this.getRequestUrl(endpoint);
 
-    const headers = new HttpHeaders( {
-      'Content-Type': 'applicaton/json'
-    })
+    return this.httpClient.get<T[]>(requestUrl, { headers: this.headers });
```
Comment (larisaChh) [inline src/app/services/http.service.ts]: is this array limitation intended? the generic type T could be User or List<User>/User[] 
So you wont need a T[] just a T that can be replace with whatever you need.

File: src/app/services/http.service.ts [hunk 3]
```diff
@@ -1,45 +1,68 @@
...
+      .set("Access-Control-Allow-Origin", "*");
+  }
+
+  private getRequestUrl(endpoint: string): string {
+    return environment.apiUrl + endpoint;
+  }
+
+  handleHttpErrorResponse(error: HttpErrorResponse): Observable<any> {
+    var errorMessage: string = `Error status ${error.status}: ${error.error}`;
+    // this.snackBar.open(errorMessage);
```
Comment (larisaChh) [inline src/app/services/http.service.ts]: you left this here for later use or useless code?

File: src/app/services/http.service.ts [hunk 4]
```diff
@@ -1,45 +1,68 @@
...
+  getById<T>(id: number, endpoint: string): Observable<T> {
+    this.setRequestHeaders();
+    const requestUrl = this.getRequestUrl(endpoint) + '/' + id;
 
+    return this.httpClient.get<T>(requestUrl, { headers: this.headers });
   }
 
-  private userPostRequest(requestUrl: string, requestBody?: any, headers?: any): Observable<any> {
-    return this.httpClient.post<User>(requestUrl, requestBody, headers);
+  post<T>(input: any, endpoint: string): Observable<T> {
```
Comment (larisaChh) [inline src/app/services/http.service.ts]: what does this "input" parameter represent? could we have a suitable name for it?

File: src/app/services/http.service.ts [hunk 5]
```diff
@@ -1,45 +1,68 @@
...
+    const requestUrl = this.getRequestUrl(endpoint);
+
+    return this.httpClient.post<T>(requestUrl, input, { headers: this.headers});
   }
 
-  private buildUrlPath(...pathParameters: string[]): string {
-    const path: string = pathParameters.reduce((previous, current) => { return previous + '/' + current });
-    
-    return environment.apiUrl + path;
+  update<T>(input: any, endpoint: string): Observable<T> {
```
Comment (larisaChh) [inline src/app/services/http.service.ts]: same as above

File: mockup-users.json [hunk 6]
```diff
@@ -1,46 +1,52 @@
 {
-  "users": [
+  "User": [
```
Comment (larisaChh) [inline mockup-users.json:2]: this object is named User but it is actually a list of users? why did you choose to change this?

Comment (larisaChh) [reply mockup-users.json:2]: also try using decent data in your mocks

**Review by larisaChh — CHANGES_REQUESTED** [review]

File: mockup-users.json (hunk 6 above)
Comment (andreibuiciuc) [reply mockup-users.json:2]: I tested the http request from my client using json-server (before trying with the real api) and from what i've seen it expects as key the endpoint (that is why I switched it to 'User', to keep it the same way I found it on the server). I didn't dive deeper into details regarding this tool, the reason was to see if the route behaves as expected.

File: src/app/services/http.service.ts (hunk 4 above)
Comment (andreibuiciuc) [reply src/app/services/http.service.ts]: Here, input represent the body of the request.
Should I use something along the line of 'requestBody' instead?

File: src/app/services/http.service.ts (hunk 3 above)
Comment (andreibuiciuc) [reply src/app/services/http.service.ts]: Useless code, my bad

File: api/Controllers/UserController.cs (hunk 1 above)
Comment (andreibuiciuc) [reply api/Controllers/UserController.cs]: Yes, it makes more sense. I'm reading about it and it is more suitable

**Review by larisaChh — APPROVED** [review]

**Pull request merged**